### PR TITLE
Simd all any none

### LINKED ISSUE #2333: Implement datapar for parallel algorithms
Here is the list of all parallel algorithms. We plan to enable most (if not all) to support `datapar`, `datapar(task)` and variations.
- [x] `adjacent_difference` (#5580)
- [x] `inner_product` (see #2330)
- [ ] `adjacent_find` 
- [ ] `all_of` `any_of` `none_of` 
- [x] `copy` `copy_n` (#5428)
- [ ] `copy_if` 
- [ ] `move`
- [x] `count` `count_if` (see #2330)
- [ ] `equal` `mismatch` 
- [ ] `exclusive_scan` `inclusive_scan` 
- [ ] `reduce` 
- [x] `transform` (see #2330)
- [x] `fill` `fill_n` (#5465)
- [ ] `find` `find_end` `find_first_of` `find_if` `find_if_not` 
- [x] `for_each` `for_each_n` (see #2330)
- [x] `generate` `generate_n` (#5594)
- [ ] `is_heap` `is_heap_until`
- [ ] `is_partitioned` `is_sorted` `is_sorted_until` 
- [ ] `lexicographical_compare` 
- [ ] `max_element` `min_element` `minmax_element` 
- [ ] `partial_sort` `partial_sort_copy` `nth_element` `sort` `stable_sort`
- [ ] `partition` `partition_copy` `stable_partition` 
- [ ] `remove` `remove_if` 
- [ ] `remove_copy` `remove_copy_if`
- [ ] `replace` `replace_copy` `replace_copy_if` `replace_if`
- [ ] `reverse` `reverse_copy`
- [ ] `rotate` `rotate_copy`
- [ ] `search` `search_n`
- [ ] `set_difference` `set_intersection` `set_symmetric_difference` `set_union` `includes`
- [ ] `inplace_merge` `merge`
- [ ] `swap_ranges`
- [ ] `transform_reduce`
- [ ] `transform_exclusive_scan` `transform_inclusive_scan`
- [ ] `uninitialized_copy` `uninitialized_copy_n`
- [ ] `uninitialized_fill` `uninitialized_fill_n`
- [ ] `unique` `unique_copy`

### PR DESCRIPTION
**Pull request by srinivasyadav18**
Contributes to solving #2333 

## Proposed Changes

  - Converts util::loop_idx_n CPO
  - Adapt new vector_pack traits
  - Adapts `all any none` to datapar
  - Adds unit tests for `all any none` to datapar

### CODE REVIEW

File: libs/core/algorithms/tests/unit/algorithms/all_of_tests.hpp [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+#include <string>
+#include <vector>
+
+#include "test_utils.hpp"
+
+///////////////////////////////////////////////////////////////////////////////
+template <typename IteratorTag>
+void test_all_of(IteratorTag)
+{
+    typedef std::vector<std::size_t>::iterator base_iterator;
```
Comment (msimberg) [inline libs/core/algorithms/tests/unit/algorithms/all_of_tests.hpp:26]:
```suggestion
    typedef std::vector<int>::iterator base_iterator;
```

and see https://github.com/STEllAR-GROUP/hpx/pull/5639. I'm hoping disabling your CircleCI builds will get us the status checks to catch this in the future.

Comment (srinivasyadav18) [reply libs/core/algorithms/tests/unit/algorithms/all_of_tests.hpp:26]: Cool, thanks!, I will do the required changes and push them soon.
And I have disabled circle ci builds for my fork.

**Comment by msimberg** [comment]
https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/9422/workflows/aed68471-ba9e-4bd5-8ea6-526c57c0930d :tada:

**Comment by srinivasyadav18** [comment]
@msimberg Hey, I think all the errors are solved, so I hope this is PR is ready. :) 
The two CIRCLE CI tests failing [here](https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/9438/workflows/c2796e9b-8676-4103-a41a-9e776c652d3f/jobs/274259) i.e `generate_datapar` and `generate_n_datapar` are solved by you in #5639 (thank you so much for that! 👍 ).

If any further changes are required, please comment!

**Review by msimberg — APPROVED** [review]
Looks good to me!

@hkaiser I'd appreciate if you could also have a closer look at this.

**Comment by hkaiser** [comment]
> Looks good to me!
> 
> @hkaiser I'd appreciate if you could also have a closer look at this.

I actually already did, just forgot to press 'Submit' on the review :/ - LGTM.

**Review by hkaiser — APPROVED** [review]

**Comment by msimberg** [comment]
bors merge

**Pull request merged**